File: 3903_smallest_stable_index_i/go/solution.go
package smallest_stable_index_i

func firstStableIndex(nums []int, k int) int {
	n := len(nums)
	suffixMin := make([]int, n)

	suffixMin[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if suffixMin[i+1] < nums[i] {
			suffixMin[i] = suffixMin[i+1]
		} else {
			suffixMin[i] = nums[i]
		}
	}

	prefixMax := nums[0]
	for i := 0; i < n; i++ {
		if nums[i] > prefixMax {
			prefixMax = nums[i]
		}
		if prefixMax-suffixMin[i] <= k {
			return i
		}
	}

	return -1
}

File: 3903_smallest_stable_index_i/kotlin/solution.kt
fun firstStableIndex(nums: IntArray, k: Int): Int {
    val suffixMin = IntArray(nums.size)
    suffixMin[nums.lastIndex] = nums.last()

    for (i in nums.size - 2 downTo 0) {
        suffixMin[i] = minOf(suffixMin[i + 1], nums[i])
    }

    var prefixMax = nums[0]
    for (i in nums.indices) {
        prefixMax = maxOf(prefixMax, nums[i])
        if (prefixMax - suffixMin[i] <= k) {
            return i
        }
    }

    return -1
}
